### PR TITLE
WT-6708 Repurpose oldest start txn to be newest txn in the aggregated time window

### PR DESCRIPTION
**Pull request by kommiharibabu**
The original intention of the oldest start transaction in the aggregated time windows
was to limit the search traversing into the page when none of the transactions is
visible to the current search snapshot. However, the aggregated oldest start transaction
doesn’t include the history store values, due to which the original intention of this
aggregated oldest start transaction doesn't limit the search. So repurposing it to store
the newest transaction will have the other benefits.

### CODE REVIEW

**Comment by luke-pearson** [comment]
Are we going to have any logic to handle the data format being changed here? As in currently doing this will mean any pre-existing WiredTiger database is going to look weird, with aggregates that don't reflect the actual lifetime of those values. If we knew that we were starting on a database that existed prior to this data format change could we revert the logic to ignore newest_start_txn and fallback to the slower implementations of truncate and RTS?

**Comment by kommiharibabu** [comment]
@luke-pearson I didn't find any change that is required to support the older data formats as the transaction ids are not valid across server restarts. The `newest_txn` value will be NONE after server start and using that value for the truncate is not a problem. It is the same behaviour even for the downgrade scenario also. As the `oldest_start_txn` will be NONE and it operates as a normal restart of the server on the same binary version. Please let me know if I missed anything that can cause problems. I ran all the compatibility and multi-version tests.

**Comment by luke-pearson** [comment]
> 
> 
> @luke-pearson I didn't find any change that is required to support the older data formats as the transaction ids are not valid across server restarts. The `newest_txn` value will be NONE after server start and using that value for the truncate is not a problem. It is the same behaviour even for the downgrade scenario also. As the `oldest_start_txn` will be NONE and it operates as a normal restart of the server on the same binary version. Please let me know if I missed anything that can cause problems. I ran all the compatibility and multi-version tests.

Good point they don't persist across restart so I think its a non issue.

File: dist/stat_data.py [hunk 1]
```diff
@@ -545,7 +545,7 @@ def __init__(self, name, desc, flags=''):
     RecStat('rec_time_aggr_newest_stop_ts', 'pages written including an aggregated newest stop timestamp '),
     RecStat('rec_time_aggr_newest_stop_txn', 'pages written including an aggregated newest stop transaction ID'),
     RecStat('rec_time_aggr_oldest_start_ts', 'pages written including an aggregated oldest start timestamp '),
-    RecStat('rec_time_aggr_oldest_start_txn', 'pages written including an aggregated oldest start transaction ID '),
+    RecStat('rec_time_aggr_newest_txn', 'pages written including an aggregated newest transaction ID '),
```
Comment (luke-pearson) [inline dist/stat_data.py:548]: I think for consistency this should be newest_start_txn, given that we use newest_start_ts and newest_start_durable_ts?

Comment (kommiharibabu) [reply dist/stat_data.py:548]: The `newest_txn` stores the maximum transaction id that is present as part of both `start transaction` and the `stop transaction` due to the aggregated `newest stop transaction` can include the `max transaction` also. Due to this reason, I didn't include the `start` as part of the name.

Comment (luke-pearson) [reply dist/stat_data.py:548]: As in I can commit a deletion which will then be the newest txn on the page?

Comment (kommiharibabu) [reply dist/stat_data.py:548]: Yes. your understanding is correct.

**Comment by kommiharibabu** [comment]
@luke-pearson, @raviprakashgiri29 and @agorrod As discussed earlier, the aggregated maximum transaction id is not only enough to skip the trees during rollback to stable. As part of WT-6746, we save the base write generation number and that is also used to skip the trees. So there is no need for any more changes that are required as part of this change.

Can you please review the code changes?

File: src/support/timestamp.c [hunk 2]
```diff
@@ -251,9 +251,9 @@ __wt_time_aggregate_validate(
           "aggregate %s",
           __wt_time_aggregate_to_string(ta, time_string[0]));
 
-    if (ta->oldest_start_txn > ta->newest_stop_txn)
+    if (ta->newest_txn > ta->newest_stop_txn)
```
Comment (luke-pearson) [inline src/support/timestamp.c:254]: Couldn't this be true if we had a newer start txn greater than a stop txn?

Comment (kommiharibabu) [reply src/support/timestamp.c:254]: No. If we have a key without a tombstone, the `newest_stop_txn` will be WT_TXN_MAX. Even if all the keys are having a tombstone, still the tombstone transaction will be the latest transaction.

File: src/include/timestamp.i [hunk 3]
```diff
@@ -104,46 +104,50 @@
...
-            (ta)->prepare = 1;                                                    \
+#define WT_TIME_AGGREGATE_UPDATE(session, ta, tw)                              \
+    do {                                                                       \
+        WT_ASSERT(session, (ta)->init_merge == 1);                             \
+        (ta)->newest_start_durable_ts =                                        \
+          WT_MAX((tw)->durable_start_ts, (ta)->newest_start_durable_ts);       \
+        (ta)->newest_stop_durable_ts =                                         \
+          WT_MAX((tw)->durable_stop_ts, (ta)->newest_stop_durable_ts);         \
+        (ta)->oldest_start_ts = WT_MIN((tw)->start_ts, (ta)->oldest_start_ts); \
+        (ta)->newest_txn = WT_MAX((tw)->start_txn, (ta)->newest_txn);          \
```
Comment (agorrod) [inline src/include/timestamp.i:126]: It deserves a comment as to why we double assign to newest_txn

Comment (kommiharibabu) [reply src/include/timestamp.i:126]: Added a comment to explain the reason for double assignment and also added some comments to the aggregated time window structure.

**Review by agorrod — APPROVED** [review]
Thanks @kommiharibabu lgtm - assuming it passes testing.

**Comment by kommiharibabu** [comment]
@luke-pearson and @raviprakashgiri29 Do you have any further comments on the PR changes?

**Comment by luke-pearson** [comment]
LGTM.

**Review by luke-pearson — APPROVED** [review]

**Review by raviprakashgiri29 — APPROVED** [review]
Changes LGTM with the context of using/renaming the oldest start txn field to store the newest start txn-id.

**Pull request merged**